[WIP] Added basic Docker support

Leopere opened the pull request:
Adding the base of a docker-compose microservices cluster for `oldboyxx/jira_clone` I need to know how to connect the Client to the API as there doesn't seem to be any documentation on any environment variables that would be used for reaching out to the API which doesn't make sense to me as an API and Client should be runnable on separate machines theoretically.  Docker best practices recommend against running more than one process per container so this design is preferred so I'm hoping that this can get fixed so I can document my changes so others can repeat this and then it can be merged.

clee commented:
Hahahaha, my first thought upon discovering this project was "I wonder if it runs in Docker yet". Was planning to try to add it myself until I found this PR.

Leopere commented:
@clee if you want to help please feel free.  I forgot to post the bloody logs I kinda rushed through this.  I do stupid little things like this to refine my skills as a DevOpserator when I'm bored.  Long live `/r/selfhosted`

Leopere commented:
```
⇒  docker-compose logs -f
Attaching to jira_clone_client_1, jira_clone_api_1, jira_clone_db_1
db_1      |
db_1      | PostgreSQL Database directory appears to contain a database; Skipping initialization
db_1      |
db_1      | LOG:  database system was shut down at 2020-01-29 00:56:10 UTC
db_1      | LOG:  MultiXact member wraparound protections are now enabled
db_1      | LOG:  database system is ready to accept connections
db_1      | LOG:  autovacuum launcher started
client_1  |
client_1  | > jira_client@1.0.0 start /app/client
client_1  | > webpack-dev-server
client_1  |
client_1  | ℹ ｢wds｣: Project is running at http://localhost:8080/
client_1  | ℹ ｢wds｣: webpack output is served from /
client_1  | ℹ ｢wds｣: Content not from webpack is served from /app/client/dev
client_1  | ℹ ｢wds｣: 404s will fallback to /index.html
client_1  | ℹ ｢wdm｣: Hash: 01951e4a726390330066
client_1  | Version: webpack 4.41.2
client_1  | Time: 18159ms
client_1  | Built at: 01/29/2020 12:56:53 AM
client_1  |                                  Asset       Size  Chunks             Chunk Names
client_1  | 029c2e0e19032f6cdbef9042552e79fa.woff2   27.4 KiB          [emitted]
client_1  |   162bf4c1112ca598251515bdd76efd60.jpg    551 KiB          [emitted]
client_1  |  1ced22ee6d45bdd05a5383fd9dbd1b43.woff   41.4 KiB          [emitted]
client_1  |  59880bb792378f77deca117ea6b17567.woff   41.3 KiB          [emitted]
client_1  |   6db192ab3ed372d1de19e207a9c46fd6.png   16.4 KiB          [emitted]
client_1  | 94a505bb1e062b41273a8ea77777a3e7.woff2   27.8 KiB          [emitted]
client_1  | 97ca95083f8be0c5d7ee9907cfe28af1.woff2   27.9 KiB          [emitted]
client_1  |  bb9a359afc17edc2c275b4212b8f2fd2.woff   40.8 KiB          [emitted]
client_1  | c8ae3e93ae16983c73a68cf537557de2.woff2   24.4 KiB          [emitted]
client_1  |  ed76eb21560f0dc038cce0adfd65ea3c.woff   33.9 KiB          [emitted]
client_1  |   f3fe29165df0323731ec27695c8d2798.svg   28.9 KiB          [emitted]
client_1  |                            favicon.png   97.3 KiB          [emitted]
client_1  |                             index.html  343 bytes          [emitted]
client_1  |                                main.js   13.3 MiB    main  [emitted]  main
client_1  | Entrypoint main = main.js
client_1  | [0] multi (webpack)-dev-server/client?http://localhost:8080 (webpack)/hot/dev-server.js ./src/index.jsx 52 bytes {main} [built]
client_1  | [./node_modules/core-js/modules/es.array-buffer.constructor.js] 563 bytes {main} [built]
client_1  | [./node_modules/core-js/modules/es.array-buffer.is-view.js] 404 bytes {main} [built]
client_1  | [./node_modules/core-js/modules/es.array-buffer.slice.js] 1.47 KiB {main} [built]
client_1  | [./node_modules/core-js/modules/es.array.concat.js] 2.34 KiB {main} [built]
client_1  | [./node_modules/core-js/modules/es.array.copy-within.js] 442 bytes {main} [built]
client_1  | [./node_modules/core-js/modules/es.array.every.js] 501 bytes {main} [built]
client_1  | [./node_modules/core-js/modules/es.array.fill.js] 399 bytes {main} [built]
client_1  | [./node_modules/core-js/modules/es.array.filter.js] 847 bytes {main} [built]
client_1  | [./node_modules/core-js/modules/es.array.find-index.js] 783 bytes {main} [built]
client_1  | [./node_modules/core-js/modules/es.array.find.js] 719 bytes {main} [built]
client_1  | [./node_modules/core-js/modules/es.array.flat-map.js] 803 bytes {main} [built]
client_1  | [./node_modules/webpack-dev-server/client/index.js?http://localhost:8080] (webpack)-dev-server/client?http://localhost:8080 4.29 KiB {main} [built]
client_1  | [./node_modules/webpack/hot/dev-server.js] (webpack)/hot/dev-server.js 1.59 KiB {main} [built]
client_1  | [./src/index.jsx] 9.5 KiB {main} [built]
client_1  |     + 983 hidden modules
client_1  | Child html-webpack-plugin for "index.html":
client_1  |      1 asset
client_1  |     Entrypoint undefined = index.html
client_1  |     [./node_modules/html-webpack-plugin/lib/loader.js!./src/index.html] 447 bytes {0} [built]
client_1  |     [./node_modules/lodash/lodash.js] 528 KiB {0} [built]
client_1  |     [./node_modules/webpack/buildin/global.js] (webpack)/buildin/global.js 472 bytes {0} [built]
client_1  |     [./node_modules/webpack/buildin/module.js] (webpack)/buildin/module.js 497 bytes {0} [built]
client_1  | ℹ ｢wdm｣: Compiled successfully.
api_1     |
api_1     | > jira_api@1.0.0 start /app/api
api_1     | > nodemon --exec ts-node --files src/index.ts
api_1     |
api_1     | [nodemon] 2.0.0
api_1     | [nodemon] to restart at any time, enter `rs`
api_1     | [nodemon] watching dir(s): *.*
api_1     | [nodemon] watching extensions: ts,json
api_1     | [nodemon] starting `ts-node --files src/index.ts`
api_1     | Wed, 29 Jan 2020 00:56:30 GMT body-parser deprecated undefined extended: provide extended option at src/index.ts:38:31
```

oldboyxx commented:
Guys, I don’t want to harsh your mellow but this is a part of my portfolio and I won’t be accepting PR’s. Feel free to fork and maintain your own version. I’ll add a note to readme.

clee commented:
Fair enough! Thanks for providing some interesting code, @oldboyxx :)

Leopere commented:
Why would portfolio work not include opensource collaboration?  Seems like a waste :(

oldboyxx commented:
Because at that point it stops being a showcase of my own skills to some degree. But most importantly, I don’t want to spend time reviewing code and dealing with PR’s on a showcase project such as this. Sorry!

Leopere commented:
That's a real shame it's effectively a dead project then?  Why would you go through the effort of making it open-source not to showcase your ability to work in an organization this seems bizarre.  I'm not trying to come off as entitled. Still, it looks like a bit of a waste because employers are going to look for your ability to collaborate, and just releasing a project not to collaborate seems like a wasted opportunity.

Leopere commented:
So not only does this not function correctly, but rather than addressing any of it, you decided to close it and openly shove people off of the project for having contributed to something that you spammed all over Reddit trying to get attention?  I hope you get a good job but my oh my.

It's not like I didn't spend a bunch of time working on this just to have the effort rejected for your portfolio.

oldboyxx commented:
Creating something in my limited spare time and sharing it with the public does not obligate me to keep spending unspecified amounts of time in perpetuity working on it because you like Docker. 

Feel free to copy the entire codebase, paste it into your own repository, and do with it what you will. You can even “spam” it on reddit to get a job.

Leopere commented:
Hey, no worries bud you're right as you clearly like skimming text I'll keep it short.  It's just reasonably considerate to consider the fact that I spent time coding something to contribute back pardon my being a bit rude about having my contribution just slammed down without even addressing the concerns I also raised in the PR.

oldboyxx commented:
Look, I realize you spent time working on this just to be rejected. I’m sorry about that. If there was an option to disable PR’s on Github I’d activate it. But for some reason there’s no such option and now we’re here arguing because of it.

I’ll add CONTRIBUTING and a note to readme.

Leopere commented:
https://help.github.com/en/github/building-a-strong-community/creating-a-pull-request-template-for-your-repository could also be of some help.

Leopere commented:
My other concern wasn't merely the Docker stuff. I was hoping that there could also be a feature, including a way to point the client at the API on a separate host. Currently, if the API has a different host IP than the Client there's no way for the client to reach out to the API

If this simple feature were added and documented, it would make this possible without me tainting your repository with my contributions.  I could even gladly be the maintainer for your appliances Docker stack for those of us hoping to be able to deploy this without breaking docker standards.  

Just a closing note: Docker isn't just because I like docker it's because it's effectively the universal software package manager; it runs across almost every platform at near-native speeds if not AT native speeds.  I was hoping to add that value to your project I don't even really care if this gets merged so long as it's added to the repository you can take all of the credit for all I care it wouldn't be the first time I just wanted to add value without needing credit, and it won't be the last.

Heck, I'd even go through and redact all of this absurd conversation if it got added. I am just trying to make the web a bit more accessible to everyone.

shirshak55 commented:
@Leopere Hey there its a MIT Licensed project so we can fork it and maintain it in our own.  I really like the project and I do have some bandwidth to maintain it so how about we maintain it? I just want to maintain not only for me but also so that it can be useful to others. And I do like your philosophy of open contribution and web accessible. Thanks.

Leopere commented:
I’m not allergic to it. Email me at jira-clone@c.nixc.us we can discuss a
fork.

On Fri, Jan 31, 2020 at 12:36 AM Shirshak <notifications@github.com> wrote:

> @Leopere <https://github.com/Leopere> Hey there its a MIT Licensed
> project so we can fork it and maintain it in our own. I really like the
> project and I do have some bandwidth to maintain it so how about we
> maintain it?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/oldboyxx/jira_clone/pull/6?email_source=notifications&email_token=AAIE2VXL6VNBUXRE33CLYLDRAO2EJA5CNFSM4KM5AHW2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKNRQVY#issuecomment-580589655>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIE2VQLLNJIFM7XRJ2CNTLRAO2EJANCNFSM4KM5AHWQ>
> .
>
